src/tests/integration/monitor-tabs.test.ts

```diff
@@ -0,0 +1,200 @@
+    })
+  })
+
+  describe('chrome.tabs.onRemoved', () => {
+    it("removes a tab from the store's state", () => {
+      const [callbackOnRemove] = mocks.chrome.tabs.onRemoved.addListener.firstCall.args
+
+      const tabId = 11
+      const removeInfo = {
+        id: 11,
+        windowId: 1,
+        active: false,
+        url: 'https://foo.com/abc',
+      }
+      callbackOnRemove(tabId, removeInfo)
+
+      const state = mocks.getState()
+      const tabs = state.tabs
+
+      expect(tabs.get(tabId)).to.equal(undefined)
+    })
+  })
+
+  describe('chrome.tabs.onUpdated', () => {
+    it("updates a tab change to store's state", () => {
+      const [callbackOnUpdated] = mocks.chrome.tabs.onUpdated.addListener.firstCall.args
+
+      const tabId = 16
+      const changeInfo = {
+        id: 16,
+        windowId: 3,
+        active: false,
+        url: 'https://changeinfo.com/abc',
+      }
+
+      callbackOnUpdated(tabId, changeInfo)
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(tabId)!
+
+      expect(tab.id).to.equal(tabId)
+      expect(tab.windowId).to.equal(changeInfo.windowId)
+      expect(tab.domain).to.equal('changeinfo.com')
+      expect(tab.url).to.equal('https://changeinfo.com/abc')
+    })
+  })
+
+  describe('chrome.tabs.onAttached', () => {
+    it("Attaches a tab to another window and updates the store's state", () => {
+      const [callbackOnAttached] = mocks.chrome.tabs.onAttached.addListener.firstCall.args
+
+      const tabId = 15
+      const attachInfo = {
+        id: 15,
+        windowId: 2,
+        newWindowId: 3,
+        active: false,
+        url: 'https://slam.com/abc',
+      }
+
+      callbackOnAttached(tabId, attachInfo)
+
+      const state = mocks.getState()
+      const tab = state.tabs.get(tabId)!
+
+      expect(tab.id).to.equal(attachInfo.id)
+      expect(tab.windowId).to.equal(attachInfo.newWindowId)
+      expect(tab.domain).to.equal('slam.com')
+      expect(tab.url).to.equal(attachInfo.url)
+    })
+  })
+
+  describe('chrome.tabs.onActivated', () => {
+    it('sets the specified tab as active for a given window and sets any other tabs of that window to be inactive', () => {
+      const [callback] = mocks.chrome.tabs.onActivated.addListener.firstCall.args
+      const activeInfo: chrome.tabs.TabActiveInfo = { tabId: 13, windowId: 1 }
+      callback(activeInfo)
+
+      const { tabs } = mocks.getState()
+      expect(tabs.get(13)).to.have.property('active', true)
+      expect(tabs.get(12)).to.have.property('active', false)
+    })
+
+    it('has no effect for active tabs of other windows', () => {
+      const { tabs } = mocks.getState()
+      expect(tabs.get(17)).to.have.property('active', true)
+    })
+  })
+
+  describe('chrome.tabs.onReplaced', () => {
+    it("replaces a tab (create and remove) and updates store's state", () => {
+      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
```

Comment:
I don't understand, this never calls the callback from `chrome.tabs.onReplaced`

Reply:
```
it("replaces a tab (create and remove) and updates store's state", () => {
      expect(mocks.chrome.tabs.onCreated.addListener).to.have.callCount(1)
      const [callbackOnCreacted] = mocks.chrome.tabs.onCreated.addListener.firstCall.args

      const addedTabId = 11
      const addedInfo = {
        id: 11,
        windowId: 1,
        active: false,
        url: 'https://foo.com/abc',
      }

      // adding previouly deleted tab
      callbackOnCreacted(addedInfo)

      // tab 18 was created from previous test
      const removedTabId = 18
      const removeInfo = {
        id: 18,
        windowId: 1,
        active: false,
        url: 'https://new.com/abc',
      }

      // console.log(removeInfo.url)

      const [callbackOnReplaced] = mocks.chrome.tabs.onReplaced.addListener.firstCall.args
      callbackOnReplaced(addedTabId, removedTabId)

      const stateOnReplaced = mocks.getState()
      const tabsNotReplaced = stateOnReplaced.tabs

      expect(tabsNotReplaced.get(removedTabId)).to.equal(undefined)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('id', addedInfo.id)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('windowId', removeInfo.windowId)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('active', removeInfo.active)
      expect(tabsNotReplaced.get(addedTabId)).to.have.property('url', removeInfo.url)
    })
```

Reply:
the Chrome.tabs.onReplace() function seems only replace the tab id, the rest of tab info are the same as deleted tab?

Reply:
Refactor as discussed.